Smooth target velocities.

eljefec opened the pull request:
* Smooth target velocities.
* Support looping through track as option.
* Estimate braking distance based on velocity at waypoint.
* Fix bug: waypoint_updater returns MAX_VELOCITY for waypoints beyond the next red light. If the car passes the stop line, then the following large velocity may pull the car through the red light.
* Remove "reduce_speed_distance" parameter.

eljefec committed:
Add conda environment.

Commit:
kinetic dockerfile

eljefec committed:
Use model trained for Nexar light detection challenge in TLDetector and TLClassifier. It has good accuracy in the simulator and a run-time between 0.02 and 0.2 seconds on my GTX 1080. It depends on Keras, TensorFlow, and OpenCV (cv2).

eljefec committed:
Use model trained for Nexar light detection challenge in TLDetector and TLClassifier. It has good accuracy in the simulator and a run-time between 0.02 and 0.2 seconds on my GTX 1080. It depends on Keras, TensorFlow, and OpenCV (cv2).

eljefec committed:
Add debug print.

eljefec committed:
Add debug print.

eljefec committed:
Add author's name for proper credit.

eljefec committed:
Add author's name for proper credit.

eljefec committed:
Experiment with running light classifier on Udacity test bags. Add scripts for generating training data for squeezeNet from test bag.

eljefec committed:
Experiment with running light classifier on Udacity test bags. Add scripts for generating training data for squeezeNet from test bag.

eljefec committed:
Add model weights. I fine-tuned challenge1.weights using a tenth of hand-labeled images from the Udacity test bags.

eljefec committed:
Add model weights. I fine-tuned challenge1.weights using a tenth of hand-labeled images from the Udacity test bags.

eljefec committed:
Add image labels for Udacity test bags.

eljefec committed:
Add image labels for Udacity test bags.

eljefec committed:
Remove unused lines.

eljefec committed:
Remove unused lines.

mwolfram committed:
merge upstream changes

mwolfram committed:
reduce rate in waypoint loader

mwolfram committed:
added required library monotonic

mwolfram committed:
added required library monotonic

Commit:
waypoint updater listens to traffic waypoint and reduces speed

eljefec committed:
Train light classifier on data collected from simulator.

eljefec committed:
Train light classifier on data collected from simulator.

eljefec committed:
Merge branch 'tl-jeff-testbag' of github.com:eva-carnd/CarND-Capstone into tl-jeff-testbag

mwolfram committed:
merge from master and implement get_closest_waypoint

eljefec committed:
Rename real to bag.

eljefec committed:
Add sim parameter to TLClassifier constructor. Class probabilities are returned by extracted function get_classification_detailed.

eljefec committed:
Add script for comparing opencv and neural-net classifiers.

mwolfram committed:
tested and fixed

mwolfram committed:
Merge branch 'master' into mw-integration

mwolfram committed:
get closest wp to traffic light

eljefec committed:
Remove dependency on monotonic because it's not available on Carla.

eljefec committed:
Remove TrainQueue from TLDetector.

eljefec committed:
Remove reference to old traffic_light_config.

eljefec committed:
Refactor: Remove commented code. Clarify member name.

eljefec committed:
Refactor: Remove unused functions.

eljefec committed:
Set up launch files for simulator and real world to select traffic light classifier.

eljefec committed:
Merge latest from tl-jeff-testbag.

mwolfram committed:
Merge pull request #5 from eva-carnd/mw-integration

Mw integration

eljefec committed:
Add team member names to README.

mwolfram committed:
name: add accents

mwolfram committed:
Merge remote-tracking branch 'upstream/master'

mwolfram committed:
add parameters

mwolfram committed:
remove processing from callback methods

mwolfram committed:
Merge pull request #6 from eva-carnd/mw-cleanup

remove processing from callback methods

mwolfram committed:
take get_closest_waypoint out of the loop

eljefec committed:
Freeze tensorflow graph for faster inference (#8)

Freeze tf graphs for squeezeNet models and incorporate frozen graphs into TLClassifier class.

eljefec committed:
Merge upstream master.

eljefec committed:
Merge pull request #9 from eva-carnd/MergeUpstream

Merge upstream

eljefec committed:
Since publisher has latch set to True, a single publish is sufficient.

eljefec committed:
Call rospy.loginfo() instead of print().

eljefec committed:
Set zero target velocity until waypoint_updater has received the waypoints and at least one traffic-waypoint message.

eljefec committed:
Lower threshold for tl_detector and increase rate of waypoint_updater.

eljefec committed:
tl_detector: Extract STATE_COUNT_THRESHOLD and rospy rate as launch parameters to allow different settings between simulator and site.

eljefec committed:
Lower warning log to info log. When tl_detector fails to start, the screen is inundated with this warning log line, which makes debugging harder.

eljefec committed:
Lower warning log to info log. When tl_detector fails to start, the screen is inundated with this warning log line, which makes debugging harder.

jfjensen committed:
fix: car drives better but fast changing tl still problematic

mwolfram committed:
reduce waypoint max velocity to 4.4m/sec

jfjensen committed:
refactor: moved some waypoint code into new function

jfjensen committed:
refactor: apply changes to be merge-ready with master

jfjensen committed:
fix: slightly better driving

eljefec committed:
Fix bug where car stops between 1st and 2nd lights. The root cause is that tl_detector detects red for a traffic light far in front, but the closest traffic light is behind the car. When waypoint_updater receives a traffic-light waypoint behind the car, it considers the car to be beside the red light and outputs a target velocity of zero. This fix works by tracking each light's shift relative to the car (i.e., getting closer or farther). This allows the traffic-light detector to ignore traffic lights behind the car.

eljefec committed:
Fix bug where car stops between 1st and 2nd lights. The root cause is that tl_detector detects red for a traffic light far in front, but the closest traffic light is behind the car. When waypoint_updater receives a traffic-light waypoint behind the car, it considers the car to be beside the red light and outputs a target velocity of zero. This fix works by tracking each light's shift relative to the car (i.e., getting closer or farther). This allows the traffic-light detector to ignore traffic lights behind the car.

eljefec committed:
tl_detector: Increase rate for simulator launch file. Stop outputting to screen (use rqt_console instead).

eljefec committed:
Decrease reduce_speed_distance. This results in a somewhat smoother slowdown near red lights. The previous value of 100 meters is rather large for slowing from 4.4 m/s to 0. The dbw_node seems to have difficulty maintaining extremely small velocities.

eljefec committed:
Check car_pose for None in tl_detector.

eljefec committed:
Check car_pose for None in tl_detector.

eljefec committed:
Merge master with jeff+jes_exp_v1.

eljefec committed:
Merge pull request #10 from eva-carnd/jeff+jes_exp_v1

Merge Jes's improvements for faster reaction to traffic lights and smoother driving

jfjensen committed:
feat: removed avoid_retention_in_follower function

jfjensen committed:
Merge pull request #11 from eva-carnd/jeff+jes_exp_v1

feat: removed avoid_retention_in_follower function

Commit:
query for light_map

eljefec committed:
Merge latest from Udacity repo, remote-tracking branch 'upstream/master'.

eljefec committed:
Refactor: Remove unused imports and unused code from tl_detector. (#12)

eljefec committed:
Change brake command to be percentage instead of torque (#14)

* feat: brake commands are now in percentages
* fix: limits between 0.8 and -0.8
* fix: carla gets values between 0.8 and -0.8

jfjensen committed:
fix: dampen throttle and brake commands significantly

jfjensen committed:
fix: adress throttle and brake issue

jfjensen committed:
fix: further improvement of PID params

jfjensen committed:
Merge pull request #15 from eva-carnd/jes-throttle

Attempt at addressing the throttle and brake issue

mwolfram committed:
merged upstream changes

mwolfram committed:
tl rate 1.0hz, add brake deadband

mwolfram committed:
styx_lo.launch for low-perfomance machines

mwolfram committed:
add brake_factor

mwolfram committed:
actually add brake_deadband, do not subtract it

eljefec committed:
Fix controller reset when DBW goes from disabled to enabled.

* It was checking dbw_enabled (the ROS msg) for None when it should check the Boolean value of dbw_enabled.data.
* I tested this in the simulator by repeatedly checking then unchecking "Manual"

mwolfram committed:
no brake_factor for Carla

mwolfram committed:
Merge pull request #17 from eva-carnd/mw-merge-wp-fixes

Mw merge wp fixes

mwolfram committed:
Merge pull request #19 from eva-carnd/mw-tl-rate

Mw tl rate

eljefec committed:
Smooth acceleration. Support track loop as option.

eljefec committed:
Remove unused function.

jfjensen reviewed (commented):
These are my comments:
1. the car stopped right on the line, rather than before the line.
2. the car didn't accelerate sufficiently when the light became green

eljefec commented:
For (1), which is better: stopping on the line or before the line?
(2) can be addressed in 2 ways: increase acceleration in waypoint_updater or tweak PID controller. However, before going down either of these routes, what do you mean by insufficient acceleration when the light turns green?

jfjensen reviewed (commented):
1. definitely before the line...
2. when the light becomes green, the car should be accelerating to move quickly through the lights... this is not a PID controller issue ... here the car should receive the max velocity to achieve the required acceleration... IMHO

mwolfram commented:
Same here: Car will not accelerate past 6mph.
I also think that the car should stop before reaching the line, then it has more time to account for slight delays.

jfjensen reviewed (changes requested):
see my comments

mwolfram reviewed (changes requested):
* The car should reach 10mph. 
* It should take minimum time to travel through the crossing. 
* The car should stop a couple of meters before reaching the stop line

eljefec committed:
Refactor: Extract Planner class so it's easier to reuse previous waypoints.

eljefec committed:
Reuse previous waypoints when appropriate.

eljefec committed:
Don't use gentle acceleration when light turns green or red light is close and car is fast.

eljefec committed:
Extract reuse and gentle to helper function. Fix bug in reuse where vi should be the velocity of the last reused waypoint.

eljefec commented:
- For stopping behind the line, I've added stop_buffer when calculating the brake distance.
- The car wasn't reaching 10mph because it kept replanning every time using the current velocity. I've updated it to reuse previously planned waypoints. Now, it reaches 10mph.
- A proper path planner should not immediately output MAX_VELOCITY; rather, it should plan a path that is safe and comfortable. If our controller was perfect, then the jerk would be unacceptably high. For master, the PID works on simulator, but there's a chance that it behaves differently on Carla and it could result in uncomfortably high acceleration.
- The project says that our acceleration should not exceed 1 m/s^2. To meet this requirement and to ensure a smooth ride on Carla, we should account for acceleration when planning target velocities. 
  - You guys were running code with acceleration of 0.25 m/s^2 (i.e., MAX_ACCEL / 4). 
  - In the latest commits, I've introduced a concept of gentleness
    - When gentle, acceleration is 0.5 m/s^2 (negative for deceleration)
    - When not gentle, acceleration is 1 m/s^2 (negative for deceleration)
  - gentle is true when:
    - A red light is far away
    - A red light is close and the car's velocity is low
  - gentle is false when:
    - A red light is close and the car's velocity is high
    - A light turns green
- This [video](https://youtu.be/Wllaw8QYM6E) is of commit 0bbc4ac running on my laptop.

eljefec committed:
Wait for traffic-light message before executing planning. Simplify Planner._get_next_velocity(). Remove unused variables.

jfjensen reviewed (changes requested):
The car missed the stop-line at the 3rd lights and it stopped in the middle of the junction. When the light became green, the car didn't move. I then manually drove the car past the junction and again turned on the DBW. The car didn't resume its course, but slowed down and came to a halt. So, there is a bug here. Probably the reference velocity is not set to max when the car is past the stop line?

eljefec commented:
UPDATE: From Jes's log below, it's clear that the target velocity is zero.

It's not clear that the bug @jfjensen experienced is due to this change. Folks on #p-system-integration and I have seen a bug in the simulator where despite full throttle, the car does not move.
https://carnd.slack.com/archives/C6NVDVAQ3/p1506794739000078
https://carnd.slack.com/archives/C6NVDVAQ3/p1506930076000089

In the bug you experienced, what was the throttle value? If the car stops moving in the simulator, make sure to check the throttle value. You can do this through `rqt_console` and filtering for "Throttle" or with command `rostopic echo /vehicle/throttle_cmd`.

jfjensen commented:
these are the values from the log:
[rosout][INFO] 2017-10-03 03:56:15,814: Velocity Ref: 0.0 - Curr: 9.0924851424e-08 - Err: -9.09248515654e-08
[rosout][INFO] 2017-10-03 03:56:15,814: Throttle Brake Steer: 0.0 0.20275844967 0.0
